### DISCUSSION TITLE
Problems with cylinder wake flow - probably multiple hard constraints

### DISCUSSION THREAD

**Discussion by BeteixZ** [lululxvi/deepxde#1195] [Q&A]
Hello, 
I'm working on a simple cylinder wake flow with the pde

$$\begin{align}
 \frac{\partial p}{\partial x}+\rho \left(v \frac{\partial u}{\partial y}+u \frac{\partial u}{\partial x}\right)-\mu \left(\frac{\partial ^2u}{\partial x^2}+\frac{\partial ^2u}{\partial y^2}\right)=0 \\
 \frac{\partial p}{\partial y}+\rho \left(u \frac{\partial v}{\partial x}+v \frac{\partial v}{\partial y}\right)-\mu \left(\frac{\partial ^2v}{\partial x^2}+\frac{\partial ^2v}{\partial y^2}\right)=0 \\
 \frac{\partial u}{\partial x}+\frac{\partial v}{\partial y}=0 \\
\end{align}$$

and boundary conditions 
- $u = 4\times 0.3y(H-y)/H^2$ and $v=0$ on the inlet,
- u=v=0 on the upper/lower and cylinder wall,
- p=0 on the outlet.

And then I write the corresponding codes as follows

```
def pde(x, uu):
    u, v, p = uu[:, 0:1], uu[:, 1:2], uu[:, 2:3]
    u_x = dde.grad.jacobian(uu, x, i=0, j=0)
    u_y = dde.grad.jacobian(uu, x, i=0, j=1)
    u_xx = dde.grad.hessian(uu, x, component=0, i=0, j=0)
    u_yy = dde.grad.hessian(uu, x, component=0, i=1, j=1)

    v_x = dde.grad.jacobian(uu, x, i=1, j=0)
    v_y = dde.grad.jacobian(uu, x, i=1, j=1)
    v_xx = dde.grad.hessian(uu, x, component=1, i=0, j=0)
    v_yy = dde.grad.hessian(uu, x, component=1, i=1, j=1)

    p_x = dde.grad.jacobian(uu, x, i=2, j=0)
    p_y = dde.grad.jacobian(uu, x, i=2, j=1)

    momentum_x = (
        rho * (u * u_x + v * u_y) - mu * (u_xx+ u_yy) + p_x
    )
    momentum_y = (
        rho * (u*v_x+v*v_y)-mu*(v_xx+ v_yy) + p_y
    )
    continuity = u_x + v_y

    return [momentum_x, momentum_y, continuity]
```

```
def inlet_u(x):
    U_mean = 0.3
    return 4 * U_mean * x[:, 1:2] * (H - x[:, 1:2]) / H ** 2
```

```
rect_domain = dde.geometry.Rectangle(xmin=[0, 0], xmax=[1.1, H])
disk_domain = dde.geometry.Disk(center=[0.2,0.2], radius= 0.05)
spatial_domain = dde.geometry.CSGDifference(rect_domain, disk_domain)

def boundary_inflow(x, on_boundary):
    return on_boundary and np.isclose(x[0], 0.0)

bc_inlet_u = dde.icbc.DirichletBC(spatial_domain, inlet_u, boundary_inflow, component=0)
```

I set up an FNN and applied the hard constraints:

```
def modify_output(xx, uu):
    x, y = xx[:, 0:1], xx[:, 1:2]
    u, v, p = uu[:, 0:1], uu[:, 1:2], uu[:, 2:3]
    u_new = y * (y - 0.41) * ((x-0.2)**2+(y-0.2)**2-0.05**2) * u
    v_new = x * y * (y - 0.41) * ((x-0.2)**2+(y-0.2)**2-0.05**2) * v
    p_new = (x - 1.1) * p
    return torch.concat((u_new, v_new, p_new), dim = 1)

data = dde.data.PDE(spatial_domain,pde,[bc_inlet_u],num_domain=10000,num_boundary=3000,num_test=10000)
net = dde.nn.FNN([2] + 4 * [50] + [3], "tanh", "Glorot normal")
net.apply_output_transform(modify_output)
```

and then trained with adam (10000 iterations) and l-bfgs (15000 iteration), the final loss is around 3e-2  basically induced by bc_inlet_u

and the result(lower) is not good compared to the ground truth(upper):
![image](https://user-images.githubusercontent.com/17287792/226172038-0154c207-01e1-4158-a838-09586f658c19.png)

I guess it's the problem of inlet_u which needs to be set as a hard constraint, but I have no idea how to couple `y * (y - 0.41) * ((x-0.2)**2+(y-0.2)**2-0.05**2) * u` with `x * u + 4*0.3*y*(H-y)/H**2`.  Can anyone help me with it? Thanks so much.

**Comment by BeteixZ** [comment]
@lululxvi Sorry to disturb you, but is it possible to set different constraints on one variable?

**Reply by lululxvi** [reply]
You could, but all hard constraints must be combined into a single one.

**Reply by BeteixZ** [reply]
Thank you, I tried multiple times to combine `y * (y - 0.41) * ((x-0.2)**2+(y-0.2)**2-0.05**2) * u` (zero u on y-bcs and disk) and `4*0.3*y*(H-y)/H**2` (inlet non-zero u on x-bc), but none of those combinations works, could you please give me some instructions? Thanks!

**Reply by lululxvi** [reply]
Then you can use soft constraints for some BCs.

**Comment by lemon123428** [comment]
@BeteixZ Hi，sir. How did you end up dealing with this? Could you show me your final code?
